Support Deequ 2.0.0 (Spark 3.1) and 2.0.1 (Spark 3.2)

lecardozo opened the pull request:
### Related issues
- #31 
- #70 
- #90 

### Related PRs
- #96 
- #75 
- #32 
 
## Changes
In order to support latest deequ releases, a few adjustments were necessary:
- [x] Added deequ **2.0.0-spark-3.1** and **2.0.1-spark-3.2** to the list of supported deequ/Spark versions
- [x] Adapted `CategoricalRangeRule` and `FractionalRangeRule` to include extra arguments for deequ>=2.0.0. 
- [x] Improved tests to ensure functionality for suggestions

Additionally, a few technical debts were addressed
- [x] Now using SparkContext to retrieve the currently installed Spark version (no need to set `SPARK_VERSION` environment variable)
- [x] Added a Github workflow to test the library with multiple Spark versions


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

lecardozo commented:
Hey, @gucciwang! 

Wanted to hear your thoughts on this PR. Let me know if there's anything else we could do to enable this support for Deequ>=2.x 😃 

Thanks!

RaptorRob commented:
Can't wait for the fix of #70

albjoaov commented:
🥺

chenliu0831 commented on .github/workflows/base.yml:
```diff
@@ -0,0 +1,48 @@
+        name: Install Python 3.6
+        if: startsWith(matrix.PYSPARK_VERSION, '2')
+        with:
+          python-version: 3.6
```
I would suggest just support Python 3.8+ officially (at least in CI), Python 3.6 is EOL already and Numpy dropped support since 1.22.0 https://numpy.org/devdocs/release/1.22.0-notes.html. There's a security issue that mandate us to upgrade to 1.22+ as well.

chenliu0831 replied:
Thanks! I would assume those workflow should start running in this parent repo once you merge, so it's ok to keep it in the same PR. Let's drop the other legacy runners in CI.

lecardozo replied:
@chenliu0831  [Here](https://github.com/lecardozo/python-deequ/actions/runs/3068883685/jobs/4956877944#step:7:224) we can see the issue happening on spark 2.4.

chenliu0831 replied:
Okay.. let's keep numpy upgrade separate for now. Does PySpark 2.4 work with python 3.8+?

lecardozo replied:
Seems good to me! I'll drop it from CI and keep numpy as 1.22+

chenliu0831 commented on poetry.lock:
```diff
@@ -437,7 +437,7 @@ name = "py4j"
 description = "Enables Python programs to dynamically access arbitrary Java objects"
 category = "main"
-optional = true
+optional = false
```
Any context on this change?

lecardozo replied:
Nice catch! I think I changed `pyspark` to become a required dependency for local development purposes, which probably made `py4j` a required dependency as well. Probably just need to re-generate the **poetry.lock** file to reflect what is defined on pyproject.toml.

Thanks for the heads up!

chenliu0831 commented on pydeequ/configs.py:
```diff
@@ -1,44 +1,36 @@
+    try:
+        return SPARK_TO_DEEQU_COORD_MAPPING[spark_version[:3]]
+    except KeyError:
+        raise RuntimeError(f"Deequ is still not supported in spark version {spark_version}")
```
Nit: let's print out supported versions? 
```suggestion
        raise RuntimeError(f"Found Incompatible Spark version {spark_version}; Use one of the Supported Spark versions for Deequ: {SPARK_TO_DEEQU_COORD_MAPPING.keys()}")
```

chenliu0831 commented on pydeequ/suggestions.py:
```diff
@@ -183,6 +185,12 @@ class CategoricalRangeRule(_RulesObject):
 
     @property
     def rule_jvm(self):
+        if IS_DEEQU_V2:
```
Hmm ideally we would like to treat V2 as the default..

lecardozo replied:
Do you mean we should drop support for <2.x and remove this condition? Or do you mean just changing the condition to something like `if IS_DEEQU_V1:` to make Deequ 1.x the exception instead of the rule?

chenliu0831 replied:
Yes, Deequ 1.x as an exception seems to be better.

chenliu0831 commented on pydeequ/suggestions.py:
```diff
@@ -183,6 +185,12 @@ class CategoricalRangeRule(_RulesObject):
     @property
     def rule_jvm(self):
+        if IS_DEEQU_V2:
+            # DISCLAIMER: this is a workaround for using the default category sorter
```
Could you shed some light why we have to pass a parameter in V2?

lecardozo replied:
Definitely! This new argument was added [here](https://github.com/awslabs/deequ/commit/5ecefaa2f2efd380e220a652be3b4eec4495cab2) on Deequ 2.x. If we try to instantiate this object without passing this parameter, py4j raises a 
```
Constructor com.amazon.deequ.suggestions.rules.CategoricalRangeRule([]) does not exist
```
exception (as reported [here](https://github.com/awslabs/python-deequ/issues/70)). The "workaround" here lives in the fact that we are forcing this parameter to the default Scala value, so we don't need to expose it to the Python API.

If you have any ideas on how we could officially support passing the `categoricalSorter` parameter from Python, I'll gladly update it! 😃

chenliu0831 replied:
Hmm I wonder how PySpark (and its Scala counterpart) generally handle those optional parameter stuff.. but it's ok for now as long as this solves #70. 

Could you open a new issue for us to revisit this?

chenliu0831 commented on tests/test_suggestions.py:
```diff
@@ -18,51 +18,75 @@
         cls.ConstraintSuggestionRunner = ConstraintSuggestionRunner(cls.spark)
         cls.sc = cls.spark.sparkContext
-        cls.df = cls.sc.parallelize([Row(a="foo", b=1, c=5), Row(a="bar", b=2, c=6), Row(a="baz", b=3, c=None)]).toDF()
+        cls.df = _create_spark_dataframe(cls.sc)
```
Ah.. we are still using vanilla unittest. Do you mind change the test runner to pytest and use [fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) for those (and probably spark session as well). It should be quite handy. 

Don't have to migrate everything, it should run `unittest.TestCases` just fine. And separate PR will be great if you like

lecardozo replied:
Yep! I can do that in a future PR to make it easier to review 😃

chenliu0831 reviewed (commented):
Awesome work @lecardozo! A few comments (mostly a bit nit) on the modernity. 

Is it possible to merge the github workflow changes first? I would love to see CI running.

chenliu0831 commented:
Is there a test report (if not from CI) that we can review? Just to make sure everything passes.

lecardozo commented:
@chenliu0831 You can check the reports of tests [here](https://github.com/lecardozo/python-deequ/actions) while I'm addressing your suggestions. As soon as we merge it, I'd expect to see them running on awslabs/pydeequ. :)

chenliu0831 commented:
> @chenliu0831 You can check the reports of tests [here](https://github.com/lecardozo/python-deequ/actions) while I'm addressing your suggestions. As soon as we merge it, I'd expect to see them running on awslabs/pydeequ. :)

Thanks, please just post a final link again when you are ready, I will review & approve

lecardozo commented:
Done! Here is the last run of the CI workflow for c410f72

chenliu0831 reviewed (approved):
🚢

mycaule commented:
Hi are you going to release the python package soon ?

matbmeijer commented:
Hi, are you expecting to give support for spark 3.2.1, and thus, databricks runtime 10.4?
